examples/utils/example-utils/src/migrationTool/sameContainerMigrationTool.ts

```diff
@@ -0,0 +1,460 @@
+	 */
+	private _v2SummaryDone: boolean = false;
+
+	private get pactMap() {
+		if (this._pactMap === undefined) {
+			throw new Error("Couldn't retrieve the PactMap");
+		}
+		return this._pactMap;
+	}
+
+	public get migrationState() {
+		// TODO: Other states
+		if (this._v2SummaryDone) {
+			return "migrated";
+			// } else if (this._finalizationStarted) {
+			// 	return "uploadingV2Summary";
+		} else if (this._quorumApprovalComplete) {
+			return "readyForMigration";
+		} else if (this._anyQuorumProposalSeen) {
+			return "waitingForV2ProposalCompletion";
+		} else if (this._v1SummaryDone) {
+			return "proposingV2Code";
+		} else if (this.acceptedVersion !== undefined) {
+			// TODO: other phases of v1 summary upload/submit?
+			return "generatingV1Summary";
+		} else if (this.proposedVersion !== undefined) {
+			return "stoppingCollaboration";
+		} else if (this._proposalP !== undefined) {
+			return "proposingMigration";
+		} else {
+			return "collaborating";
+		}
+	}
+
+	// Once we have the v2 contents in hand, we can work on sending it as the next summary (which will look like a v2
+	// summary, due to the v2 code details being in the quorum).  This is the final step.
+	// TODO: Real param typing
+	public async finalizeMigration(v2Summary: string) {
+		// TODO: Start by awaiting connected?
+		// If someone else finishes this while the local client is still working on it, the local
+		// client should immediately abort this step and move on to the next one.
+		// Retain the summary handle for retry in subsequent phases
+		// TODO: actual implementation
+		// TODO: This is basically executing the summary flow outside of the container.  How much of this should stay inside the container?
+		// Main concern is that we want to race it with the summary already being done by another client, and not retry after it's done.
+		// TODO: We may want all clients to race to avoid the problem of the missing v2 summaryAck?  BUT need to acknowledge that sucessful upload
+		// is only "done" for single-commit summary.
+
+		// Don't allow premature calls.
+		if (!this._quorumApprovalComplete) {
+			throw new Error("Not ready to finalize until quorum approval has completed");
+		}
+		// TODO: Also guard against repeat calls
+		const uploadV2Summary = async (_v2Summary) => {
+			// Do upload of v2Summary
+			// TODO: Retry also
+			this.emit("uploadingV2Summary");
+			console.log(_v2Summary);
+			return "This is the handle I got back after successfully uploading the v2 summary";
+		};
+		const submitV2Summary = async (_v2SummaryHandle) => {
+			// Submit summarize op for v2Summary
+			this.emit("submittingV2Summary");
+			// TODO: Retry also
+			// TODO: Returning a never-resolving promise for now so we at least wait on some real summary.
+			return new Promise((resolve) => {});
+		};
+		if (this._v2SummaryDone) {
+			return;
+		}
+		const v2SummaryHandle = await Promise.race([uploadV2Summary(v2Summary), this._v2SummaryP]);
```

Comment:
i didn't know about `Promise.race()`, this is cool 🙂

Reply:
Yep!  Ultimately this might go away after thinking through failure/retry modes (e.g. we'll want to abort the v2 summary submission if we see someone else do it first, which might change this bit).  But it at least hints at the idea that we aren't the only one trying to submit the summary here and we're OK with someone else beating us to it.